### PR TITLE
Fix edge case compatibility in `cupy.eye()`

### LINKED ISSUE #5704: Edge cases of compatibility in `numpy.eye()`
These are discovered when running hypothesis-based Array API tests. For now I am just documenting these discoveries. I do not know if they are something we'd like to fix -- I'd need more time to look into them. 

Case 1:
```python
>>> np.eye(0, k=9223372036854775808, dtype=np.int8)
array([], shape=(0, 0), dtype=int8)
>>>
>>> cp.eye(0, k=9223372036854775808, dtype=np.int8)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leof/dev/cupy_cuda114/cupy/_creation/basic.py", line 114, in eye
    ret.diagonal(k)[:] = 1
  File "cupy/_core/core.pyx", line 873, in cupy._core.core.ndarray.diagonal
    cpdef ndarray diagonal(self, offset=0, axis1=0, axis2=1):
  File "cupy/_core/core.pyx", line 881, in cupy._core.core.ndarray.diagonal
    return _indexing._ndarray_diagonal(self, offset, axis1, axis2)
  File "cupy/_core/_routines_indexing.pyx", line 218, in cupy._core._routines_indexing._ndarray_diagonal
    return _diagonal(self, offset, axis1, axis2)
OverflowError: Python int too large to convert to C ssize_t
```

Case 2:
```python
>>> np.eye(1, k=-9223372036854775808, dtype=np.int8)
array([[0]], dtype=int8)
>>>
>>> cp.eye(1, k=-9223372036854775808, dtype=np.int8)
array([[1]], dtype=int8)
```

### PR DESCRIPTION
**Pull request by takagi**
Fix #5704

### CODE REVIEW

File: cupy/_creation/basic.py [hunk 1]
```diff
@@ -110,7 +110,11 @@ def eye(N, M=None, k=0, dtype=float, order='C'):
     """
     if M is None:
         M = N
-    ret = zeros((N, M), dtype, order=order)
+    ret = zeros((N, M), dtype=dtype, order=order)
+    if k >= M:
+        return ret
+    if k <= -N:
+        return ret
```
Comment (emcastillo) [inline cupy/_creation/basic.py:117]:
```suggestion
    if k <= -N or k >= M:
        return ret
```
stupid nit :)

**Review by toslunar — COMMENTED** [review]
The simple change to `eye` LGTM and I approve to merge the PR as it is, while it looks better to completely rewrite `_ndarray_diagonal` in `cupy/_core/_routines_indexing.pyx` in order to fix diag methods, too.

File: tests/cupy_tests/creation_tests/test_basic.py [hunk 2]
```diff
@@ -172,11 +169,12 @@ def test_empty_zero_sized_array_strides(self, order):
         b = cupy.empty((1, 0, 2), dtype='d', order=order)
         assert b.strides == a.strides
 
+    @pytest.mark.parametrize('offset', [1, -1, 1<<63, -(1<<63)])
```
Comment (emcastillo) [inline tests/cupy_tests/creation_tests/test_basic.py:172]:
```suggestion
    @pytest.mark.parametrize('offset', [1, -1, 1 << 63, -(1 << 63)])
```
Flake8 😅

Comment (takagi) [reply tests/cupy_tests/creation_tests/test_basic.py]: Thanks!

**Review by toslunar — APPROVED** [review]
LGTM

Thanks @takagi and @emcastillo.

**Pull request merged**